File: backend/src/utils/accessCode.js
const crypto = require("crypto");

const ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";

function generateSegment(length = 4) {
  let result = "";

  for (let i = 0; i < length; i++) {
    result += ALPHABET[crypto.randomInt(0, ALPHABET.length)];
  }

  return result;
}

function generateAccessCode() {
  return `MV-${generateSegment(4)}-${generateSegment(4)}`;
}

function hashAccessCode(accessCode) {
  const normalizedCode = accessCode.trim().toUpperCase();
  const salt = crypto.randomBytes(16);

  const derivedKey = crypto.scryptSync(
    normalizedCode,
    salt,
    64
  );

  return `scrypt$${salt.toString("hex")}$${derivedKey.toString("hex")}`;
}

function verifyAccessCode(accessCode, storedHash) {
  if (!accessCode || !storedHash) {
    return false;
  }

  try {
    const normalizedCode = accessCode.trim().toUpperCase();

    const parts = storedHash.split("$");

    if (parts.length !== 3 || parts[0] !== "scrypt") {
      return false;
    }

    const salt = Buffer.from(parts[1], "hex");
    const storedKey = Buffer.from(parts[2], "hex");

    const derivedKey = crypto.scryptSync(
      normalizedCode,
      salt,
      storedKey.length
    );

    return crypto.timingSafeEqual(derivedKey, storedKey);
  } catch (error) {
    return false;
  }
}

function isValidAccessCodeFormat(accessCode) {
  if (typeof accessCode !== "string") {
    return false;
  }

  return /^MV-[A-Z0-9]{4}-[A-Z0-9]{4}$/.test(
    accessCode.trim().toUpperCase()
  );
}

module.exports = {
  generateAccessCode,
  hashAccessCode,
  verifyAccessCode,
  isValidAccessCodeFormat,
};

File: backend/app/utils/access_code.py
import hashlib
import hmac
import re
import secrets

ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"


def generate_access_code() -> str:
    segment = lambda: "".join(secrets.choice(ALPHABET) for _ in range(4))
    return f"MV-{segment()}-{segment()}"


def hash_access_code(access_code: str) -> str:
    salt = secrets.token_bytes(16)
    derived = hashlib.scrypt(access_code.strip().upper().encode(), salt=salt, n=16384, r=8, p=1, dklen=64)
    return f"scrypt${salt.hex()}${derived.hex()}"


def verify_access_code(access_code: str, stored_hash: str | None) -> bool:
    if not access_code or not stored_hash:
        return False
    try:
        prefix, salt_hex, key_hex = stored_hash.split("$")
        if prefix != "scrypt":
            return False
        stored_key = bytes.fromhex(key_hex)
        derived = hashlib.scrypt(access_code.strip().upper().encode(), salt=bytes.fromhex(salt_hex), n=16384, r=8, p=1, dklen=len(stored_key))
        return hmac.compare_digest(derived, stored_key)
    except (ValueError, TypeError):
        return False


def is_valid_access_code_format(access_code: object) -> bool:
    return isinstance(access_code, str) and bool(re.fullmatch(r"MV-[A-Z0-9]{4}-[A-Z0-9]{4}", access_code.strip().upper()))
